### ISSUE TITLE
Addition service load balancer disruption on azure while upgrading from 1.23 to 1.23

### ISSUE THREAD

**Issue by deads2k** [kubernetes/kubernetes#108112]
### What happened?

Doing a micro upgrade from one level of 1.23 to another level of 1.23, we noticed disruption to  PDB based app went from about 5s to about 13s when compared to a 1.22.z to 1.22.z+1 update.

### What did you expect to happen?

I did not expect any disruption.

### How can we reproduce it (as minimally and precisely as possible)?

This one is a bit difficult, we have thousands of runs to compare averages and distributions against.  It's specifically azure, it's specifically the service load balancer, it's worst on 1.23 to 1.23 upgrades versus 1.22 to 1.23 upgrades.

### Anything else we need to know?

We tracked it down to one of these three commits:

1. https://github.com/kubernetes/kubernetes/pull/105253
2. https://github.com/kubernetes/kubernetes/pull/105839
3. https://github.com/kubernetes/kubernetes/pull/105188

When we revert all three, we see improved behavior (our test was in https://github.com/openshift/kubernetes/pull/1148).

We attempted to workaround by setting the exclude-from-lb label, but this filter prevents the change from being observed by the informer's cache: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/azure/azure.go#L752-L755

When we attempted to remove the filter entirely (attempt here https://github.com/openshift/kubernetes/pull/1174/files#diff-c94e623481fb506d90b33e51830aabc3e8829a308a70dc9e4ebd0f31d436edc1R779), we get
> I0212 12:57:01.695808       1 azure_standard.go:1005] EnsureBackendPoolDeleted: skipping the node ci-ln-87q96z2-1d09d-hnn5d-worker-westus-5x7n9 belonging to another vm set CI-LN-87Q96Z2-1D09D-HNN5D_CI-LN-87Q96Z2-1D09D-HNN5D-WORKER-WESTUS-AS 

from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/azure/azure_standard.go#L1003-L1007

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

Azure

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
ping @nilo19 @feyskyer

**Comment by ricky-rav** [comment]
Adding more details to the issue... 

We analyzed how Azure Load Balancer backends are modified as cluster nodes become or stop being available, which in our tests can happen in two ways:

- by labeling nodes with `node.kubernetes.io/exclude-from-external-load-balancers`

- by rebooting nodes

With kubernetes prior to this PR (https://github.com/kubernetes/kubernetes/pull/105839, solving https://github.com/kubernetes-sigs/cloud-provider-azure/issues/851), namely kubernetes 1.22, shipped with openshift 4.9, we notice that:

1. the cloud service controller in `staging/src/k8s.io/cloud-provider/controllers/service/controller.go` receives event notifications on nodes and updates accordingly the list of available nodes for load balancers (https://github.com/openshift/kubernetes/blob/b93fd35dd030519c24dd02f8bc2a7f873d2928cd/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L668-L688): a node is excluded from a load balancer backend pool if it has the "exclude" label above, or if its status is not `Ready`;

2. `reconcileLoadBalancer` in `azure_loadbalancer.go` receives the updated list of nodes from the cloud service controller
   
3. if this list differs from the current list of nodes in the backend pool for a given load balancer, it removes spurious nodes and adds any new nodes.
   
4. in particular, node deletion begins here: https://github.com/openshift/kubernetes/blob/release-4.9/staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go#L1137-L1162 and takes place in the `EnsureBackendPoolDeleted` function.

So far so good, if I label a node that I want to exclude from a load balancer, the following logs appear:

> I0211 14:38:44.651083       1 azure_loadbalancer.go:1142] reconcileLoadBalancer for service (openshift-ingress/router-default)(true): lb backendpool - found unwanted node ci-ln-3qh25qk-1d09d-5m9q8-worker-centralus2-zg6hc, decouple it from
> the LB
> I0211 14:38:44.852854       1 azure_backoff.go:90] GetVirtualMachineWithRetry(ci-ln-3qh25qk-1d09d-5m9q8-worker-centralus2-zg6hc): backoff success
> I0211 14:38:44.881964       1 azure_standard.go:1031] EnsureBackendPoolDeleted begins to CreateOrUpdate for NIC(ci-ln-3qh25qk-1d09d-5m9q8-rg, ci-ln-3qh25qk-1d09d-5m9q8-worker-centralus2-zg6hc-nic) with backendPoolID
> /subscriptions/d38f1e38-4bed-438e-b227-833f997adf6a/resourceGroups/ci-ln-3qh25qk-1d09d-5m9q8-rg/providers/Microsoft.Network/loadBalancers/ci-ln-3qh25qk-1d09d-5m9q8/backendAddressPools/ci-ln-3qh25qk-1d09d-5m9q8
> 

To confirm that the deletion took place on the Azure side, the output of `az network lb list -g $myRG` taken before and after adding the above label shows clearly that the affected node was removed from the backend pool of this load balancer.

Now, if we try to do the same with kubernetes 1.23 (planned to ship with openshift 4.10), we see that the above deletion NEVER takes place:
- steps 1 and 2 from above are the same: the cloud service controller realizes that a node has been labeled (or rebooted) and removes from the list of nodes it passes to the Azure `econcileLoadBalancer`
- The PR mentioned above, however, changes the logic of how nodes are removed and the result is that deletion never happens, at least for the two cases mentioned above (a node that gets the "exclude" label, a node that is rebooting).


From what I understand in the current code:
- deletion now happens if this call returns true: `az.ShouldNodeExcludedFromLoadBalancer(nodeName)`
- this function does NOT make use of the node list received as input by the reconciler. It looks for the node name in `az.excludeLoadBalancerNodes`, which is updated by `az.updateNodeCaches`, which is in turn triggered by separate node informers in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/azure/azure.go#L741-L777
- for some reason, these node informers trigger a cache update only if the topology label (`v1.LabelTopologyZone`) on them changes, so the list in `az.excludeLoadBalancerNodes` is only correct at startup, and then becomes stale. As a consequence, in our tests, as nodes reboot or get labeled, the code always checks for inclusion against an empty `az.excludeLoadBalancerNodes` list...

**The update logic with these node informers was not part of the PR mentioned above. Is this something that was done on purpose?**

In order to overcome this, I tried removing the condition on the topology label and triggering a cache update on every update event.
This goes a little further, but deletion is skipped because `EnsureBackendPoolDeleted` only issues a delete if 

> I0212 12:57:03.499252       1 azure_standard.go:1005] EnsureBackendPoolDeleted: skipping the node ci-ln-87q96z2-1d09d-hnn5d-worker-westus-5x7n9 belonging to another vm set CI-LN-87Q96Z2-1D09D-HNN5D_CI-LN-87Q96Z2-1D09D-HNN5D-WORKER-WESTUS-AS

Indeed, if I add debug lines I see that:

- `vmasName=CI-LN-87Q96Z2-1D09D-HNN5D_CI-LN-87Q96Z2-1D09D-HNN5D-WORKER-WESTUS-AS`

- `vmSet=385791c7-1c0a-4d8c-a6a5-d01104521d42`

I am not familiar with these terms and what they entail. What's interesting, though, is that this code hasn't changed all that much since openshift 4.9, so I added debug lines to see what the values for `vmasName` and `vmSet` were back then and it turned out that they were both empty strings! The check for equality was then successful and deletion took place.

Now, why were they both empty?
`vmSet` is an input parameter in `EnsureBackendPoolDeleted` and is set here, in the reconciler function:

>     vmSetName := az.mapLoadBalancerNameToVMSet(lbName, clusterName)
>     err = az.VMSet.EnsureBackendPoolDeleted(service, lbBackendPoolID, vmSetName, backendpoolToBeDeleted)

In one of my tests, `lbName` was for instance `ci-ln-stbgb3b-1d09d-2xxfd`
(unfortunately I did not log clusterName...)

```
    func (az *Cloud) mapLoadBalancerNameToVMSet(lbName string, clusterName string) (vmSetName string) {
	    vmSetName = strings.TrimSuffix(lbName, InternalLoadBalancerNameSuffix) // removing -internal suffix
	    if strings.EqualFold(clusterName, vmSetName) {
		    vmSetName = az.VMSet.GetPrimaryVMSetName()
	    }
	    return vmSetName
    }
```

My understanding is that `az.VMSet.GetPrimaryVMSetName()` returns an empty value in 4.9:

```
// GetPrimaryVMSetName returns the VM set name depending on the configured vmType.
// It returns config.PrimaryScaleSetName for vmss and config.PrimaryAvailabilitySetName for standard vmType.
func (as *availabilitySet) GetPrimaryVMSetName() string {
        return as.Config.PrimaryAvailabilitySetName
}
```


This code wasn't modified in the PR or since 1.22, so I do not see how now with 1.23 we finally get a proper value for `vmSetName`...


So to recap, 

- The node cache updates take place only when the topology label changes. Was this done on purpose? I don't see it mentioned in the issue linked to the original PR

- even if we trigger updates to the node cache systematically (disregarding the topology label check),  we then hit another problem that prevents the delete operation to take place. This involves some subtleties in the Azure configuration (vmasName vs vmSet, and also Config.PrimaryAvailabilitySetName) and we definitely need someone from the upstream community to take a look at any code changes we might introduce in there...

**Comment by ricky-rav** [comment]
(I just updated my long comment above with more precise info on openshift 4.9)

**Comment by bridgetkromhout** [comment]
Thanks, @deads2k & @ricky-rav! We'll try to get the right folks looking at this.

**Comment by vpickard** [comment]
Thanks @bridgetkromhout , appreciate all your help and efforts on this one!

**Comment by nilo19** [comment]
@ricky-rav thanks for catching this, I will file a PR to fix it ASAP. By the way, is it possible to switch to out-of-tree cloud provider azure since it has been GA?

**Comment by ricky-rav** [comment]
The fix would be rather simple, something along the lines of https://github.com/openshift/kubernetes/pull/1190
What do you guys think? Thanks!
